2D context get/putImageData cache should be aware of empty contents

kkinnunen-apple opened the pull request:
#### abc23583c3fded932dce57b3a4cdfa547d771464
<pre>
2D context get/putImageData cache should be aware of empty contents
<a href="https://bugs.webkit.org/show_bug.cgi?id=264930">https://bugs.webkit.org/show_bug.cgi?id=264930</a>
<a href="https://rdar.apple.com/118499949">rdar://118499949</a>

Reviewed by Cameron McCormack.

Avoid real ImageBuffer::getPixelBuffer when getImageData is called
for a new 2D context. Track the transparent contents state and
just return zero filled ImageData buffer.

Hardens ImageData to ensure the data buffer is not created too big.
The buffer size is limited with PixelBuffer max size.
This would be caught with test
fast/canvas/canvas-getImageData-invalid-result-buffer-crash.html

Fixes a bug where CG path of text drawing would not mark the canvas
as modified. This would be caught with test
fast/canvas/gradient-text-with-shadow.html

Fixes a bug where focus drawing would not mark the canvas as modified.
This would be caught with test
fast/canvas/draw-focus-if-needed-with-path.html
fast/canvas/draw-focus-if-needed.html

* Source/WebCore/html/ImageData.cpp:
(WebCore::ImageData::create):
* Source/WebCore/html/ImageData.h:
* Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
(WebCore::CanvasRenderingContext2DBase::reset):
(WebCore::CanvasRenderingContext2DBase::didDraw):
(WebCore::CanvasRenderingContext2DBase::evictCachedImageData):
(WebCore::CanvasRenderingContext2DBase::CachedContentsImageData::CachedContentsImageData):
(WebCore::CanvasRenderingContext2DBase::cacheImageDataIfPossible):
(WebCore::CanvasRenderingContext2DBase::makeImageDataIfContentsCached const):
(WebCore::CanvasRenderingContext2DBase::getImageData const):
(WebCore::CanvasRenderingContext2DBase::putImageData):
(WebCore::CanvasRenderingContext2DBase::CachedImageData::CachedImageData): Deleted.
(WebCore::CanvasRenderingContext2DBase::takeCachedImageDataIfPossible const): Deleted.
* Source/WebCore/html/canvas/CanvasRenderingContext2DBase.h:

Canonical link: <a href="https://commits.webkit.org/271462@main">https://commits.webkit.org/271462@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2762e2504380bc9dcbf27d8874172e799ed181ac

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/27879 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/6518 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/29154 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/30410 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/25446 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/28373 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/8494 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/3912 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/25220 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/28145 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/5275 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/23934 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/4551 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/4734 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/24949 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/31098 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/25481 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/25392 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/30943 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/4743 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/2898 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/28809 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/6264 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6820 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/5189 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/5211 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

kkinnunen-apple commented on Source/WebCore/html/canvas/CanvasRenderingContext2DBase.h:
```diff
@@ -362,12 +335,41 @@ class CanvasRenderingContext2DBase : public CanvasRenderingContext, public Canva
     void didDraw(bool entireCanvas, const FloatRect&, OptionSet<DidDrawOption> options = defaultDidDrawOptions());
     template<typename RectProvider> void didDraw(bool entireCanvas, RectProvider, OptionSet<DidDrawOption> options = defaultDidDrawOptions());
 
+    static String normalizeSpaces(const String&);
```
These seem to move but they're just because didDraw and needed things have to move to protected, since there are derived classes that draw.

heycam commented on Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
```diff
@@ -2366,42 +2365,52 @@ RefPtr<ByteArrayPixelBuffer> CanvasRenderingContext2DBase::cacheImageDataIfPossi
-    convertImagePixels(source, destination, size);
-    return ImageData::create(size, WTFMove(data), m_settings.colorSpace);
+    return WTF::switchOn(m_cachedContents,
+        [&](CachedContentsTransparent&)  -> RefPtr<ImageData> {
```
Nit: one too many spaces before `->`.

heycam commented on Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
```diff
@@ -2366,42 +2365,52 @@ RefPtr<ByteArrayPixelBuffer> CanvasRenderingContext2DBase::cacheImageDataIfPossi
+        [&](CachedContentsImageData& cachedImageData) -> RefPtr<ImageData> {
+            // Always consume the cache.
+            Ref pixelBuffer = WTFMove(cachedImageData.imageData);
+            m_cachedContents.emplace<CachedContentsUnknown>();
```
I always find this kind of pattern a bit worrying, since it leaves us with a dangling `CachedContentsImageData& cachedImageData`. I don't have a good suggestion on what to do instead, unless we change to use something other than `WTF::switchOn`.

kkinnunen-apple replied:
What about the ifs in the new patch

heycam commented on Source/WebCore/platform/graphics/PixelBuffer.cpp:
```diff
@@ -48,13 +48,16 @@ bool PixelBuffer::supportedPixelFormat(PixelFormat pixelFormat)
+    ASSERT_UNUSED(pixelFormat, supportedPixelFormat(pixelFormat));
 
     constexpr unsigned bytesPerPixel = 4;
+    auto bufferSize = CheckedUint32 { size.width() } * size.height() * bytesPerPixel;
+    if (!bufferSize.hasOverflowed() && bufferSize.value() > std::numeric_limits<int32_t>::max())
+        bufferSize.overflowed();
+    return bufferSize;
```
Why do we need to ensure the computed size fits within an `int32_t` too? I see that the caller does that check currently, but what relies on that?

FWIW you could also do this with:

```
CheckedUInt32 PixelBuffer::computeBufferSize(...)
{
    ASSERT_UNUSED(...);

    constexpr unsigned bytesPerPixel = 4;
    return CheckedInt32 { CheckedUInt32 { size.width() } * size.height() * bytesPerPixel };
}
```

if you think that's just as clear. (I assume you moved away from `area()` since that also supports computing sizes involving negative dimensions.)

kkinnunen-apple replied:
> I see that the caller does that check currently, but what relies on that?

 I suppose it's a succint way of saying width > 0 && height > 0. Some ImageData(-5, 5) kind of tests fail without it.

Your proposal without clarifying comment is easy to mistake for a mistake: "Somebody forgot one U here ".

How about writing it explicitly like in the patch?

kkinnunen-apple replied:
Scrap that, [fast/canvas/canvas-getImageData-invalid-result-buffer-crash.html](https://github.com/WebKit/WebKit/blob/main/LayoutTests/fast/canvas/canvas-getImageData-invalid-result-buffer-crash.html) fails with naive iffing

kkinnunen-apple committed:
2D context get/putImageData cache should be aware of empty contents
https://bugs.webkit.org/show_bug.cgi?id=264930
rdar://118499949

Reviewed by Cameron McCormack.

Avoid real ImageBuffer::getPixelBuffer when getImageData is called
for a new 2D context. Track the transparent contents state and
just return zero filled ImageData buffer.

Hardens ImageData to ensure the data buffer is not created too big.
The buffer size is limited with PixelBuffer max size.
This would be caught with test
fast/canvas/canvas-getImageData-invalid-result-buffer-crash.html

Fixes a bug where CG path of text drawing would not mark the canvas
as modified. This would be caught with test
fast/canvas/gradient-text-with-shadow.html

Fixes a bug where focus drawing would not mark the canvas as modified.
This would be caught with test
fast/canvas/draw-focus-if-needed-with-path.html
fast/canvas/draw-focus-if-needed.html

* Source/WebCore/html/ImageData.cpp:
(WebCore::ImageData::create):
* Source/WebCore/html/ImageData.h:
* Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
(WebCore::CanvasRenderingContext2DBase::reset):
(WebCore::CanvasRenderingContext2DBase::didDraw):
(WebCore::CanvasRenderingContext2DBase::evictCachedImageData):
(WebCore::CanvasRenderingContext2DBase::CachedContentsImageData::CachedContentsImageData):
(WebCore::CanvasRenderingContext2DBase::cacheImageDataIfPossible):
(WebCore::CanvasRenderingContext2DBase::makeImageDataIfContentsCached const):
(WebCore::CanvasRenderingContext2DBase::getImageData const):
(WebCore::CanvasRenderingContext2DBase::putImageData):
(WebCore::CanvasRenderingContext2DBase::CachedImageData::CachedImageData): Deleted.
(WebCore::CanvasRenderingContext2DBase::takeCachedImageDataIfPossible const): Deleted.
* Source/WebCore/html/canvas/CanvasRenderingContext2DBase.h:

Canonical link: https://commits.webkit.org/271462@main
